### ISSUE TITLE
Specs are unintuitive

### ISSUE THREAD

**Issue by effigies** [nipype/pydra#428]
Working with latest `master`...

I'm trying to add an `update_spec(spec, dict)` function that would make it easier to separate initialization and argument population and solve issues like #417 by simply making it so `pydra.mark.task` doesn't pass inputs to `FunctionTask()` but instead updates its `.inputs` structure. (For some reason attrs doesn't provide in-place evolution...)

In trying to write a test for this, I ran into the issue that making a spec without all the wrapping of a `Task` is painful.

```Python
import attr
from pydra.engine.helpers import make_klass
from pydra.engine.specs import BaseSpec, SpecInfo

TestSpec = make_klass(
    SpecInfo(
        name="TestSpec",
        fields=[("a", attr.ib(type=int)), ("b", attr.ib(type=str))],
        bases=(BaseSpec,)
    )
)

TestSpec()
```

Results in:

```
TypeError: __init__() missing 2 required keyword-only arguments: 'a' and 'b'
```

I can, however, write:

```Python
TestSpec(a=attr.NOTHING, b=attr.NOTHING)
```

Can we make this more manageable? In particular, initializing with no arguments should produce a spec with values that are all `attr.NOTHING` or some class-specified default value.

**Comment by satra** [comment]
`a = attr.evolve(a, ...)` so not inplace but that's the pattern for attr.

regarding initialization both attrs and dataclasses tries to not allow initialization  without specifying something. so one has to follow this pattern:

https://github.com/nipype/pydra/blob/master/pydra/engine/core.py#L134

regarding the issue in #417 `func` is a reserved keyword for the Function Task. we could technically call it anything, but another attrs behavior is that `_func` maps to `func` (so `getattr(obj, 'func')` will return the hidden attribute). i believe there is an issue about this in pydra somewhere. there is no concept of a private field in attrs (they are quite opinionated about this).  so we would have to come up with something that is unlikely to be used by a function.

**Comment by satra** [comment]
just extending the above code to map onto the pattern.

```python
In [4]: inputs = TestSpec( 
   ...:             **{ 
   ...:                 # in attrs names that starts with "_" could be set when name provided w/o "_" 
   ...:                 (f.name[1:] if f.name.startswith("_") else f.name): f.default 
   ...:                 for f in attr.fields(TestSpec) 
   ...:             } 
   ...:         )                                                                                                   

In [5]: inputs                                                                                                      
Out[5]: TestSpec(a=NOTHING, b=NOTHING)

```

**Comment by satra** [comment]
we could wrap the above pattern into a function:

```python

def init_klass(klass):
    return klass(
            **{
                # in attrs names that starts with "_" could be set when name provided w/o "_"
                (f.name[1:] if f.name.startswith("_") else f.name): f.default
                for f in attr.fields(klass)
            }
        )
```

**Comment by djarecka** [comment]
btw. @effigies suggested today that perhaps we can use `__func` instead of `_func`, but it looks like it doesn't help. `attr.make_class` doesn't like when I pass a dictionary with `func` and `__func`

**Comment by satra** [comment]
you could consider one of the following patterns. the first one automatically enters the class name into field, so the dictionary would have to contain `<classname>__func`. the second would make it explicit but would still not allow that matching key.

```python
In [23]: @attr.s 
    ...: class Foo: 
    ...:     __func = attr.ib(type=str) 
    ...:                                                                                                            

In [24]: foo = Foo(Foo__func=1)                                                                                     

In [25]: foo._Foo__func                                                                                             
Out[25]: 1

In [26]: @attr.s 
    ...: class Foo: 
    ...:     private__func = attr.ib(type=str) 
    ...:                                                                                                            

In [27]: foo = Foo(private__func=1)                                                                                 

In [28]: foo.private__func                                                                                          
Out[28]: 1
```

**Comment by effigies** [comment]
> `a = attr.evolve(a, ...)` so not inplace but that's the pattern for attr.

Yes, I did see this. So would `task.inputs = attr.evolve(task.inputs, ...)` be kosher? I guess there shouldn't be any hidden state that's worth preserving, but is there any chance that someone could hold onto a stale copy of `task.inputs` and fail to get the updates?

> regarding initialization both attrs and dataclasses tries to not allow initialization without specifying something.

Might need refreshing on why we are using `attr.NOTHING` instead of `None` to indicate no value. I guess because it's a legitimate Python value. Maybe it's worth creating our own `pydra.NOTHING` or `pydra.UNDEFINED`?

I need to think through the rest of this thread, but I wanted to consider posting an issue to attrs to see if our use case can be supported upstream: https://gist.github.com/effigies/b8474891962bfe55ca114d4a296a20bf

Need to do some other things now, so posting the draft for your thoughts before posting to attrs.

**Comment by djarecka** [comment]
@effigies - yes, we're using `evolve` in this way, e.g. https://github.com/nipype/pydra/blob/master/pydra/engine/core.py#L430

And we didn't want to use `None` on purpose, to be able to use it as a "valid" value for the field.

**Comment by effigies** [comment]
I think I would be up for `__func` -> `_FunctionTask__func`.
